src/dune_rules/exe_rules.ml

```diff
@@ -122,57 +131,55 @@ let executables_rules ~sctx ~dir ~expander ~dir_contents ~scope ~compile_info
     in
     let dynlink =
       (* See https://github.com/ocaml/dune/issues/2527 *)
-      (* XXX rgrinberg: what is this? *)
-      true
-      || Dune_file.Executables.Link_mode.Map.existsi exes.modes
-           ~f:(fun mode _loc ->
-             match mode with
-             | Other { kind = Shared_object; _ } -> true
-             | _ -> false)
+      Dune_file.Executables.Link_mode.Map.existsi exes.modes
```

Comment:
Unfortunately we need to keep `dynlink` to be `true` for now because the `-nodynlink` mode of the compiler is buggy (it generates invalid code in certain architectures).

Reply:
Good call. I reverted.